kernel/arch/dreamcast/hardware/g2dma.c

```diff
@@ -0,0 +1,248 @@
+
+    top_range = (value & 0x7f00) >> 8;
+    bottom_range = (value & 0x7f);
+
+    top_addr = (top_range << 20) | 0x08000000;
+    bottom_addr = (bottom_range << 20) | 0x080fffff;
+*/
+#define SYS_MEM_SECURITY_CODE 0x4659
+#define DISABLE_SYS_MEM_PROTECTION (SYS_MEM_SECURITY_CODE << 16 | 0x007F)
+#define ENABLE_SYS_MEM_PROTECTION  (SYS_MEM_SECURITY_CODE << 16 | 0x7F00)
+
+/*  
+    Set the DS# (Data Strobe) timeout to 27 clock cycles for the external DMA. 
+    If data isn't ready for latching by this time, an interrupt will be 
+    triggered. 
+    
+    Not sure why its 27 but can be changed later. Default value 
+    is 1023 cycles (0x3ff).
+*/
+#define DS_CYCLE_OVERRIDE  27
+
+/* Disable the DMA */
+inline static void dma_disable(uint32_t chn) {
+    g2_dma->dma[chn].enable = 0;
+    g2_dma->dma[chn].start = 0;
+}
+
+static void g2_dma_irq_hnd(uint32_t code) {
+    int chn = code - ASIC_EVT_G2_DMA0;
+
+    if(chn < SPU_DMA_G2CHN || chn > CH3_DMA_G2CHN) {
+        dbglog(DBG_ERROR, "g2_dma: Wrong channel received in g2_dma_irq_hnd");
+        return;
+    }
+
+    /* VP : changed the order of things so that we can chain dma calls */
+
+    /* Signal the calling thread to continue, if any. */
+    if(dma_blocking[chn]) {
+        sem_signal(&dma_done[chn]);
+        thd_schedule(1, 0);
+        dma_blocking[chn] = 0;
+    }
+
+    /* Call the callback, if any. */
+    if(dma_callback[chn]) {
+        dma_callback[chn](dma_cbdata[chn]);
+    }
+}
+
+int g2_dma_transfer(void *sh4, void *g2bus, size_t length, uint32_t block,
+                    g2_dma_callback_t callback, void *cbdata,
+                    uint32_t dir, uint32_t mode, uint32_t g2chn, uint32_t sh4chn) {
+    /* No longer used but we keep then around for compatibility */
+    (void)mode;
+    (void)sh4chn;
+
+    if(g2chn > CH3_DMA_G2CHN) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    /* Check alignments */
+    if(((uint32_t)sh4) & 31) {
+        dbglog(DBG_ERROR, "g2_dma: Unaligned sh4 DMA %p\n", sh4);
+        errno = EFAULT;
+        return -1;
+    }
+
+    if(((uint32_t)g2bus) & 31) {
+        dbglog(DBG_ERROR, "g2_dma: Unaligned g2bus DMA %p\n", g2bus);
+        errno = EFAULT;
+        return -1;
+    }
+
+    /* Make sure length is a multiple of 32 */
+    length = (length + 0x1f) & ~0x1f;
+
+    dma_blocking[g2chn] = block;
+    dma_callback[g2chn] = callback;
+    dma_cbdata[g2chn] = cbdata;
+
+    /* Make sure we're not already DMA'ing */
+    if(g2_dma->dma[g2chn].start != 0) {
+        dbglog(DBG_ERROR, "g2_dma: Already DMA'ing for channel %ld\n", g2chn);
+        errno = EINPROGRESS;
+        return -1;
+    }
+
+    /* Set needed registers */
+    g2_dma->dma[g2chn].g2_addr = ((uint32_t)g2bus) & MASK_ADDRESS;
+    g2_dma->dma[g2chn].sh4_addr = ((uint32_t)sh4) & MASK_ADDRESS;
+    g2_dma->dma[g2chn].size = length | RESET_ENABLED;
+    g2_dma->dma[g2chn].dir = dir;
+    g2_dma->dma[g2chn].trigger_select = CPU_TRIGGER;
```

Comment:
@DC-SWAT See here. We currently disable suspend and there is no issues setting the suspend register even if it is suspended. I need to enable suspend though and test that out

Reply:
I think we should enable suspending here.